### PR TITLE
Performance: dequeue jobs in batches

### PR DESCRIPTION
**Pull request by hkdsun**
Yesterday, during an incident that affected cloud pods, we observed resque-scheduler causing job scheduling SLO violation.

image

We tracked down the violation to the high volume of webhook jobs that were being scheduled for the future (put into the resque-scheduler queue as opposed to resque). Shedding this load solved the unbounded queue growth.

This PR is a first attempt at addressing this performance problem by batching the operation of dequeuing jobs. The most important caveat of this solution is that the number of jobs that _can_ be dropped by a hard crash is increased from a single job to `dequeue_batch_size` jobs. Furthermore, since jobs are dequeued in batches, a failure in Redis that lasts through the period of enqueueing a batch of jobs is likely to affect the whole batch.

I profiled the dequeueing + enqueueing process and found that we spend an equal amount of time  dequeuing as we do enqueueing with a batch size of 1:
batch-1

I then profiled the code with a batch size of 10:
batch10

As you can see, we are still spending a considerable amount of time _enqueueing_ jobs into Redis. The enqueue codepath is more complicated to batch up and possibly not worth the effort for now.

Below is benchmark results (benchmark code in the diff) that show the impact of dequeue batching (n.b. this is a local Redis instance. Results in production may vary). I picked a compromise between speedup and damage from failures and used a batch size of 10 by default (tuneable in an env variable):
image

As mentioned, the results in the production environment may vary due to different latency conditions.

@Shopify/core-cloud-swat @Shopify/redis-memcached

### CODE REVIEW

**Comment by xldenis** [comment]
Thanks for this PR @hkdsun, there are a couple things I think we should consider first. 

- Could we accelerate things without batching by using lua scripts? ie: avoiding network transfer entirely? That would also help protect us during degraded network conditions. 
- Why do you feel like 10 is more appropriate than 5? I feel like doubling our safety loss for 6% speed up isn't necessarily the best default.

File: lib/resque/scheduler/delaying_extensions.rb [hunk 1]
```diff
@@ -68,13 +68,10 @@ def delayed_queue_schedule_size
...
       end
 
-      def next_delayed_item(before:)
-        item, time = redis.zrangebyscore(:delayed_queue, 0.0, before.to_i, limit: [0, 1], with_scores: true).first
-
-        if item
-          redis.zrem(:delayed_queue, item)
-          decode_without_nonce(item)
-        end
+      def next_delayed_item(before:, count: 1)
```
Comment (xldenis) [inline lib/resque/scheduler/delaying_extensions.rb]: could we pipeline this?

Comment (xldenis) [inline lib/resque/scheduler/delaying_extensions.rb]: can we rename this to `next_delayed_items`?

File: lib/resque/scheduler/configuration.rb [hunk 2]
```diff
@@ -68,6 +68,14 @@ def poll_sleep_amount
         @poll_sleep_amount ||=
           Float(ENV.fetch('RESQUE_SCHEDULER_INTERVAL', '5'))
       end
+
+      # Dequeueing batch size
+      attr_writer :dequeue_batch_size
+
+      def dequeue_batch_size
+        @dequeue_batch_size ||=
+          Integer(ENV.fetch('RESQUE_SCHEDULER_DEQUEUE_BATCH_SIZE', '10'))
```
Comment (xldenis) [inline lib/resque/scheduler/configuration.rb]: see my comment, but I think 5 would be a better default here.

File: lib/resque/scheduler.rb [hunk 3]
```diff
@@ -177,21 +177,22 @@ def env_matches?(configured_env)
...
-              if item
+              items.each do |item|
                 log "queuing #{item['class']} [delayed]"
                 enqueue(item)
               end
             end
           end
 
-          break if item.nil?
+          break if !master || items.empty?
```
Comment (csfrancis) [inline lib/resque/scheduler.rb]: Why did you add this check? Wouldn't the loop break on the next iteration?

File: lib/resque/scheduler/configuration.rb [hunk 4]
```diff
@@ -68,6 +68,14 @@ def poll_sleep_amount
         @poll_sleep_amount ||=
           Float(ENV.fetch('RESQUE_SCHEDULER_INTERVAL', '5'))
       end
+
+      # Dequeueing batch size
```
Comment (csfrancis) [inline lib/resque/scheduler/configuration.rb]: Nitpick on the needless comment :-)

File: lib/resque/scheduler/delaying_extensions.rb [hunk 5]
```diff
@@ -68,13 +68,10 @@ def delayed_queue_schedule_size
...
+        items = redis.zrangebyscore(:delayed_queue, 0.0, before.to_i, limit: [0, count], with_scores: false)
```
Comment (csfrancis) [inline lib/resque/scheduler/delaying_extensions.rb]: `zrangebyscore` will always return an array?

File: lib/resque/scheduler/delaying_extensions.rb (hunk 1 above)
Comment (csfrancis) [inline lib/resque/scheduler/delaying_extensions.rb]: Perhaps the method name should now be plural, since it's returning an array of items?

Comment (csfrancis) [reply lib/resque/scheduler/delaying_extensions.rb]: OR maybe have a helper `next_delayed_item` that returns the first element from `next_delayed_items`.

File: lib/resque/scheduler.rb (hunk 3 above)
Comment (hkdsun) [reply lib/resque/scheduler.rb]: This is a subtle bug I found while testing. 

Say the instance loses master status after assigning `items` to something non-empty (in L186), then we would never break out of the loop AFAICT.

Btw, I thought about changing L185 to be `break unless master?` but that only breaks out of the `handle_shutdown` block, not the loop

**Comment by hkdsun** [comment]
> Could we accelerate things without batching by using lua scripts? ie: avoiding network transfer entirely? That would also help protect us during degraded network conditions. If we could figure out a way to handle processing one job entirely in lua we'd likely get the best speed up there.

As you mention, that would require a rethinking of the enqueue logic, which will be a bit sensitive because the enqueue logic is delegated to the application: https://github.com/Shopify/shopify/blob/e3ef5466c9e4b8ac03c74ad94ea656d9b85104ce/eagerlib/background_queue/recurring_job.rb#L11

I think the changes here are simpler, keep resque-scheduler generic, and still give it a good speedup.

> Why do you feel like 10 is more appropriate than 5? I feel like doubling our safety loss for 6% speed up isn't necessarily the best default.

I have no strong opinion – I think in general we drop very small number of jobs and I thought in a higher latency network a batch of 10 might line up better with my benchmarking. 

For failure stats, see the "Scheduler: # Failures to Enqueue for DC and pod" in the [Resque dashboard](https://shopify.datadoghq.com/dash/73626/resque?live=true&page=0&is_auto=false&from_ts=1518617888070&to_ts=1518621488070&tile_size=m).

Here is the metric over the past month for the whole platform (side note: we should investigate why those are failing more often):

image

> Also, I'd like to hear your opinion on the enqueue rewrite. I have some as well ;)

I haven't thought about this much but I think we should start by modifying the core resque code to be able to enqueue batches of jobs.

**Comment by pushrax** [comment]
One of the biggest benefits I see in using Lua for this would be that it could perform the operation atomically. Right now, as far as I can tell, resque-scheduler is another source of "AtLeastOnce actually means AtLeastNever"

**Comment by hkdsun** [comment]
>Right now, as far as I can tell, resque-scheduler is another source of "AtLeastOnce actually means AtLeastNever"

We don't drop a whole lot of jobs from resque-scheduler. See my graph in the last post for this stat over the past month.

> How much of a yak shave would it be to change the interaction in some way to make that possible?

We would essentially be bypassing anything that happens in: https://github.com/resque/resque/blob/a6eed6d06e5f8d743dea492c147f0c1191118adb/lib/resque.rb#L365-L379

This includes [some enqueue hooks](https://github.com/resque/resque/blob/a6eed6d06e5f8d743dea492c147f0c1191118adb/lib/resque.rb#L367-L376) and [a simple validation](https://github.com/resque/resque/blob/a6eed6d06e5f8d743dea492c147f0c1191118adb/lib/resque/job.rb#L88)

I think the most yak shavey part would be banning `{before,after,around}_enqueue` hooks and clearing the shitlist of all instances of it (which doesn't seem that much but still few days of work probably).

I don't think I have time to go down that rabbit hole but maybe @Shopify/redis-memcached would? 

Alternatively, we can start with this fix and revisit if we see failures happen often (I can own looking at the metric once a week or so)

**Comment by xldenis** [comment]
We could use a lua script that copies the batch to a temp list before passing it down to resque scheduler. We could delete it after the enqueue.

File: test/delayed_queue_test.rb [hunk 6]
```diff
@@ -109,14 +109,28 @@
...
+    assert_equal 10, Resque::Scheduler.dequeue_batch_size
+    30.times { Resque.enqueue_at(t, SomeIvarJob) }
+
+    # 30 SomeIvarJob jobs should be created in the "ivar" queue
+    Resque::Job.expects(:create).times(30).with('ivar', SomeIvarJob, nil)
     Resque::Scheduler.handle_delayed_items(t)
   end
 
+  test 'handle_delayed_items uses batch size of 10 by default' do
+    assert_equal 10, Resque::Scheduler.dequeue_batch_size
```
Comment (fbogsany) [inline test/delayed_queue_test.rb]: Why assert this rather than stubbing it?

File: test/delayed_queue_test.rb [hunk 7]
```diff
@@ -109,14 +109,28 @@
 
   test 'handle_delayed_items with items in the future' do
     t = Time.now + 60 # in the future
-    Resque.enqueue_at(t, SomeIvarJob)
-    Resque.enqueue_at(t, SomeIvarJob)
 
-    # 2 SomeIvarJob jobs should be created in the "ivar" queue
-    Resque::Job.expects(:create).twice.with('ivar', SomeIvarJob, nil)
+    assert_equal 10, Resque::Scheduler.dequeue_batch_size
```
Comment (fbogsany) [inline test/delayed_queue_test.rb]: Does this test care about the `dequeue_batch_size`?

Comment (hkdsun) [reply test/delayed_queue_test.rb]: Good point. This is bad expression of my idea for this test. 

The intention is to test `handle_delayed_items` when there are more jobs in the queue than the batch size. I'll make it more clear by using a `batch_size` variable and doing `(batch_size * 3).times { Resque.enqueue_at(t, SomeIvarJob) }`

File: test/delayed_queue_test.rb (hunk 6 above)
Comment (fbogsany) [reply test/delayed_queue_test.rb]: :+1: should have read the name of the test 😞
